lib/lz4frame.c

```diff
@@ -188,7 +202,15 @@ typedef struct LZ4F_cctx_s
     U64    totalInSize;
     XXH32_state_t xxh;
     void*  lz4CtxPtr;
-    U32    lz4CtxLevel;   /* 0: unallocated;  1: LZ4_stream_t;  3: LZ4_streamHC_t */
+    /* lz4CtxLevel records both what size of memory has been allocated into the
+     * ctx pointer field and what it's being used as.
+     * - The low bit (& 1) value indicates whether the ctx is being used for
+     *   hc (1) or not (0).
+     * - The next bit (& 2) indicates whether the allocated memory is big
+     *   enough for a non-hc context.
+     * - The next bit (& 4) indicates whether the allocated memory is big
```

Comment:
Don't we have `&4` -> implies `&2` ?

Reply:
Yes. I can add a comment to that effect.

Reply:
Or, you could use an assignment scheme which makes it impossible to generate a bogus statement by accident, such as "large enough for hc, but not large enough for fast".

What about a 3-states value `{0,1,2}` ?

Reply:
ping

Reply:
Also : mixing flag bits into a pointer is a near trick to save memory, which can be handy when the structure size is constrained.
But it doesn't seem to be the case here.
In contrast with LZ4 context type,
`struct LZ4F_cctx_s` is an incomplete type from a public API perspective,
so there is no way to allocate it statically,
so there is no structure size to preserve.

It's likely better for clarity to have dedicated field names.

Reply:
Ok, yeah, if we don't care about continuity of struct layout, I'll split `lz4CtxLevel` into two fields, one indicating the allocated size and one indicating what it's being used for now.